packages/dataviews/src/dataform-controls/array.tsx

```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
-	// Ensure value is an array
-	const arrayValue = useMemo(
+	// Convert stored values to element objects for the token field
+	const arrayValueAsElements = useMemo(
 		() =>
 			Array.isArray( value )
 				? value.map( ( token ) => {
-						const tokenLabel = findElementByValue( token )?.label;
-						return tokenLabel || token;
+						const element = elements?.find(
+							( suggestion ) => suggestion.value === token
+						);
+						return element || { value: token, label: token };
 				  } )
 				: [],
-		[ value, findElementByValue ]
+		[ value, elements ]
 	);
 
-	const onChangeControl = useCallback(
-		( tokens: ( string | { value: string } )[] ) => {
-			// Convert TokenItem objects to strings
-			const stringTokens = tokens.map( ( token ) => {
-				if ( typeof token !== 'string' ) {
+	const validateTokens = useCallback(
+		( tokens: ( string | { value: string; label?: string } )[] ) => {
+			// Extract actual values from tokens for validation
+			const tokenValues = tokens.map( ( token ) => {
+				if ( typeof token === 'object' && 'value' in token ) {
 					return token.value;
 				}
+				return token;
+			} );
+
+			// First, check if elements validation is required and any tokens are invalid
+			if ( field.isValid?.elements && elements ) {
+				const invalidTokens = tokenValues.filter( ( tokenValue ) => {
+					return ! elements.some(
+						( element ) => element.value === tokenValue
+					);
+				} );
+
+				if ( invalidTokens.length > 0 ) {
+					setCustomValidity( {
+						type: 'invalid',
+						message: __(
+							'Please select from the available options.'
+						),
+					} );
+					return;
+				}
+			}
+
+			// Then check custom validation if provided
+			if ( field.isValid?.custom ) {
+				const result = field.isValid.custom(
+					{
+						...data,
+						[ id ]: tokenValues,
+					},
+					field
+				);
+
+				if ( result ) {
+					setCustomValidity( {
+						type: 'invalid',
+						message: result,
+					} );
+					return;
+				}
+			}
 
-				const tokenByLabel = findElementByLabel( token );
+			// If no validation errors, clear custom validity
+			setCustomValidity( undefined );
+		},
+		[ elements, data, id, field ]
+	);
 
-				return tokenByLabel?.value || token;
+	// Generate custom invalid message for elements validation
+	const customInvalidMessage = useMemo( () => {
+		if ( field.isValid?.elements && elements ) {
+			return __( 'Please select from the available options.' );
+		}
+		return __( 'Invalid item' );
+	}, [ field.isValid?.elements, elements ] );
+
+	const onChangeControl = useCallback(
+		( tokens: ( string | { value: string; label?: string } )[] ) => {
+			const valueTokens = tokens.map( ( token ) => {
+				if ( typeof token === 'object' && 'value' in token ) {
+					return token.value;
+				}
+				// If it's a string, it's either a new suggestion value or user input
+				return token;
 			} );
 
 			onChange( {
-				[ id ]: stringTokens,
+				[ id ]: valueTokens,
 			} );
+
+			validateTokens( tokens );
```

Comment:
`onValidate` is called every time `onChange` is called, right? So this would trigger validation twice.

Reply:
Great catch! Indeed, it was running twice. I don't think we need it here, so I removed it 08f9b87d26c7d795b4603602724d17a2dcf8839b